### ISSUE TITLE
SDXL lora traning weird result

### ISSUE THREAD

**Issue by Dudekpob** [tdrussell/diffusion-pipe#162]
Hello i train sdxl lora like 7k steps character lora and got weird result almost identical parameters like in kohya_ss but withour cosine_with_restart 
can someone help how to get better result??
Repeat =25 
images = 30+ 

```

save_every_n_epochs = 4  #
epochs = 10
micro_batch_size_per_gpu = 2 gradient_accumulation_steps = 1
#gradient_release=true
activation_checkpointing = true
#blocks_to_swap = 2  
dataset = 'dataset.toml'


[model]
type = 'sdxl'
checkpoint_path = 'models/sdxl/matureRitualMILFOFGOD_onepro.safetensors'
dtype = 'bfloat16'
# You can train v-prediction models (e.g. NoobAI vpred) by setting this option.
#v_pred = true
# Min SNR is supported. Same meaning as sd-scripts
#min_snr_gamma = 5
# Debiased estimation loss is supported. Same meaning as sd-scripts.
#debiased_estimation_loss = true
# You can set separate learning rates for unet and text encoders. If one of these isn't set, the optimizer learning rate will apply.
unet_lr = 1e-4
text_encoder_1_lr = 1e-4
text_encoder_2_lr = 1e-4

[adapter]
type = "lora"  
rank = 8  
#dropout = 0.1  
dtype = "bfloat16"  
# init_from_existing = "path/to/pretrained/lora"  


[optimizer]
type = "adamw8bit" 
lr = 0.0002  # Learning rate
#gradient_clipping = 1.0  


```

**Comment by tdrussell** [comment]
What does "weird" mean? Is the quality just a bit lower, or did things get completely fried?

When you trained with sd-scripts, did you train text encoders? Because your config shows you training text encoders with what is IMO a pretty high learning rate.

I can tell you that I did a large training run on SDXL using diffusion-pipe, with completely identical settings as a lora I trained with sd-scripts. I compared results with X/Y grid comparisons and the images were nearly identical between the two for each seed.

Also if you can make your dataset a bit larger, splitting off a validation dataset and computing validation loss is always a good idea to tell you if the model is learning in a generalizable manner and when it starts to overfit.

**Comment by tdrussell** [comment]
```
[optimizer]
type = "adamw8bit" 
```
Wait, this might affect things. Not sure though. You should use adamw_optimi, or AdamW8bitKahan (look at the example config file). diffusion-pipe uses full bf16 training, and relies on the optimizer having Kahan summation so that bf16 roundoff error isn't a concern. sd-scripts typically uses mixed precision so the weights and optimizer states are in float32. You could also just set the lora dtype to float32 instead.

Or maybe none of that actually matters, it's just something I noticed.

**Comment by Dudekpob** [comment]
got some better result


```
output_dir = "runs/lora_training" 
save_every_n_epochs = 4  
epochs = 10  
micro_batch_size_per_gpu = 2  
gradient_accumulation_steps = 1  
#gradient_release=true
activation_checkpointing = true
#blocks_to_swap = 2 
dataset = 'dataset.toml'

[model]
type = 'sdxl'
checkpoint_path = 'models/sdxl/matureRitualMILFOFGOD_onepro.safetensors'
dtype = 'bfloat16'
# You can train v-prediction models (e.g. NoobAI vpred) by setting this option.
#v_pred = true
# Min SNR is supported. Same meaning as sd-scripts
#min_snr_gamma = 5
# Debiased estimation loss is supported. Same meaning as sd-scripts.
#debiased_estimation_loss = true
# You can set separate learning rates for unet and text encoders. If one of these isn't set, the optimizer learning rate will apply.
unet_lr = 1e-4
text_encoder_1_lr = 1e-4
text_encoder_2_lr = 1e-4


[adapter]
type = "lora"  
rank = 8  
#dropout = 0.1  
dtype = "bfloat16"  
# init_from_existing = "path/to/pretrained/lora"  



[optimizer]
type = "adamw8bit"  
lr = 0.0002  # Learning rate
#gradient_clipping = 1.0  

```

**Comment by lowcaloriesyrup** [comment]
@tdrussell - How can we compute validation loss? Is this done during normal training, by pointing to a separate validation dataset? Or does it need to be done independent of training?

**Comment by tdrussell** [comment]
You have a separate validation config file. The example TOML file shows how to do this.
